_overviews/scala3-book/domain-modeling-tools.md

```diff
@@ -0,0 +1,575 @@
+- Having an `unapply` method auto-generated for you also lets case classes be used in advanced ways with pattern matching. (They take a little longer to demonstrate, so they’re shown in detail in the Reference documentation.)
+
+As they write in the book, [Programming in Scala](https://www.amazon.com/Programming-Scala-Updated-2-12/dp/0981531687/) (Odersky, Spoon, and Venners), “the biggest advantage of case classes is that they support pattern matching.”
+
+
+{% comment %}
+We can use this following text, if desired. It might need to be updated a little.
+
+### An `unapply` method
+
+In the previous lesson on companion objects you saw how to write `unapply` methods. A great thing about a case class is that it automatically generates an `unapply` method for your class, so you don’t have to write one.
+
+To demonstrate this, imagine that you have this trait:
+
+```scala
+trait Person {
+    def name: String
+}
+```
+
+Then, create these case classes to extend that trait:
+
+```scala
+case class Student(name: String, year: Int) extends Person
+case class Teacher(name: String, specialty: String) extends Person
+```
+
+Because those are defined as case classes — and they have built-in `unapply` methods — you can write a match expression like this:
+
+```scala
+def getPrintableString(p: Person): String = p match {
+    case Student(name, year) =>
+        s"$name is a student in Year $year."
+    case Teacher(name, whatTheyTeach) =>
+        s"$name teaches $whatTheyTeach."
+}
+```
+
+Notice these two patterns in the `case` statements:
+
+```scala
+case Student(name, year) =>
+case Teacher(name, whatTheyTeach) =>
+```
+
+Those patterns work because `Student` and `Teacher` are defined as case classes that have `unapply` methods whose type signature conforms to a certain standard. Technically, the specific type of pattern matching shown in these examples is known as a *constructor pattern*.
+
+>The Scala standard is that an `unapply` method returns the case class constructor fields in a tuple that’s wrapped in an `Option`. The “tuple” part of the solution was shown in the previous lesson.
+
+To show how that code works, create an instance of `Student` and `Teacher`:
+
+```scala
+val s = Student("Al", 1)
+val t = Teacher("Bob Donnan", "Mathematics")
+```
+
+Next, this is what the output looks like in the REPL when you call `getPrintableString` with those two instances:
+
+```scala
+scala> getPrintableString(s)
+res0: String = Al is a student in Year 1.
+
+scala> getPrintableString(t)
+res1: String = Bob Donnan teaches Mathematics.
+```
+
+>All of this content on `unapply` methods and extractors is a little advanced for an introductory book like this, but because case classes are an important FP topic, it seems better to cover them, rather than skipping over them.
+{% endcomment %}
+
```

Comment:
You could show how to add pattern matching to any type by implementing `unapply`

Reply:
I added an `unapply` example. This text is actually inside a comment block. Would you (all) prefer that I remove the comments and make it part of the page?